src/uint.rs

```diff
@@ -152,42 +144,76 @@ macro_rules! uint_overflowing_add {
 		(U512(result), overflow != 0)
 	});
 
-	($name:ident, $n_words:expr, $self_expr: expr, $other: expr) => (
+	($name:ident, $n_words:tt, $self_expr: expr, $other: expr) => (
 		uint_overflowing_add_reg!($name, $n_words, $self_expr, $other)
 	)
 }
 
 #[cfg(not(all(asm_available, target_arch="x86_64")))]
 macro_rules! uint_overflowing_sub {
-	($name:ident, $n_words: expr, $self_expr: expr, $other: expr) => ({
+	($name:ident, $n_words:tt, $self_expr: expr, $other: expr) => ({
 		uint_overflowing_sub_reg!($name, $n_words, $self_expr, $other)
 	})
 }
 
-macro_rules! uint_overflowing_sub_reg {
-	($name:ident, $n_words: expr, $self_expr: expr, $other: expr) => ({
+macro_rules! uint_overflowing_binop {
+	($name:ident, $n_words:tt, $self_expr: expr, $other: expr, $fn:expr) => ({
 		let $name(ref me) = $self_expr;
 		let $name(ref you) = $other;
 
-		let mut ret = [0u64; $n_words];
+		let mut ret = unsafe { ::core::mem::uninitialized() };
+		let ret_ptr = &mut ret as *mut [u64; $n_words] as *mut u64;
 		let mut carry = 0u64;
 
-		for i in 0..$n_words {
-			let (res1, overflow1) = me[i].overflowing_sub(you[i]);
-			let (res2, overflow2) = res1.overflowing_sub(carry);
+		unroll! {
+			for i in 0..$n_words {
+				use ::core::ptr;
 
-			ret[i] = res2;
-			carry = overflow1 as u64 + overflow2 as u64;
+				if carry != 0 {
```

Comment:
why do you need to check carry here?

why not just 

```
let (res1, overflow1) = ($fn)(me[i], you[i]);
let (res2, overflow2) = ($fn)(res1, carry);

<ptr::write...>

let carry = overflow1 | overflow2
```

since overflow can happen only once?